### PR TITLE
ciao-image: Add support for private, public and internal images

### PR DESCRIPTION
**Pull request by obedmr**
It's using a new metadata field called ``visibility`` which defines the access
restrictions that users and CIAO inself have when trying to access the images
catalog. Below are the 3 main types of visibility that are supported:

- Internal
      - CIAO cluster's internal images (e.g. cnci)
      - Only created by members of admin tenant/project
- Private
      - Default visibility for any created image
      - Only available in the tenant/project where it was created
- Public
      - Only created/deleted by members of admin tenant/project
      - Available in all tenants/projects

### CODE REVIEW

**Comment by markdryan** [comment]
@obedmr You need to update singleVM to set the visibility of the standard images it creates, otherwise it cannot create any of the default workloads.

@albertomurillo Do we create images in the ansible scripts with ciao-cli image add?  If so, these will need to be uploaded as well.

File: ciao-controller/openstack_image.go [hunk 1]
```diff
@@ -80,10 +85,10 @@ func (is ImageService) CreateImage(req image.CreateImageRequest) (image.DefaultR
 	return image.DefaultResponse{
 		Status:     image.Queued,
 		CreatedAt:  i.CreateTime,
-		Tags:       make([]string, 0),
+		Tags:       strings.Split(i.Tags, " "),
```
Comment (markdryan) [inline ciao-controller/openstack_image.go]: Yes, I think I'm correct

```
Image #3
	Name             [Clear Linux 12620]
	Size             [907018240 bytes]
	UUID             [df3768da-31f5-4ba6-82f0-127a1a705169]
	Status           [active]
	Visibility       [public]
	Tags             [[]]
	CreatedDate      [2017-02-16T17:37:30.561028069Z]
```

Doesn't look right.

**Comment by markdryan** [comment]
@obed What are the rules about creation and deletion of public images?

In your comments above you say that they can only be created by admin.  However, I was able to create one in single VM as the normal user.

Also who can delete public images?  Currently, it only seems to be the user that created the images.  Admin can see public images but he cannot delete them.  If he tries he gets a weird 404 error.

File: ciao-image/datastore/metads.go [hunk 2]
```diff
@@ -27,39 +28,51 @@ type MetaDs struct {
 
 // Write is the metadata write implementation.
 func (m *MetaDs) Write(i Image) error {
-	err := m.DbAdd("images", i.ID, &i)
+	tenant := i.TenantID
+	if i.Visibility == imageAPI.Public {
+		tenant = string(imageAPI.Public)
```
Comment (markdryan) [inline ciao-image/datastore/metads.go]: Should this be

image.Public

?

File: ciao-image/datastore/metads.go [hunk 3]
```diff
@@ -27,39 +28,51 @@ type MetaDs struct {
...
+	return m.DbDelete(tenant, id)
 }
 
 // Get is the metadata get implementation.
-func (m *MetaDs) Get(ID string) (Image, error) {
+func (m *MetaDs) Get(tenant, ID string) (Image, error) {
 
 	imageTable := &ImageMap{}
-	img, err := m.DbGet("images", ID, imageTable)
+	img, err := m.DbGet(string(imageAPI.Public), ID, imageTable)
```
Comment (markdryan) [inline ciao-image/datastore/metads.go]: Again should this be image.Public?

File: ciao-cli/image.go [hunk 4]
```diff
@@ -65,6 +71,8 @@ func (cmd *imageAddCommand) parseArgs(args []string) []string {
 	cmd.Flag.StringVar(&cmd.id, "id", "", "Image UUID")
 	cmd.Flag.StringVar(&cmd.file, "file", "", "Image file to upload")
 	cmd.Flag.StringVar(&cmd.template, "f", "", "Template used to format output")
+	cmd.Flag.BoolVar(&cmd.public, "public", false, "Public Image")
```
Comment (markdryan) [inline ciao-cli/image.go]: I see, internal is overloaded.  It's used to indicate an internal image and for tags.  Is there any reason why we can't have tags on a public or private image?

File: ciao-cli/image.go [hunk 5]
```diff
@@ -89,9 +97,26 @@ func (cmd *imageAddCommand) run(args []string) error {
...
+
+	imageVisibility := images.ImageVisibilityPrivate
+	if cmd.public {
+		imageVisibility = images.ImageVisibilityPublic
+	}
+
+	tags := []string{}
+	if cmd.internal != "" {
+		imageVisibility = internalImage
+		tags = append(tags, cmd.internal)
```
Comment (markdryan) [inline ciao-cli/image.go]: Shouldn't we split the string?

File: ciao-cli/image.go (hunk 4 above)
Comment (obedmr) [inline ciao-cli/image.go]: @markdryan, agree, I just wanted to use tags for putting internal images information. But agree, let's allow public and private images have tags.
Regarding visibility flag, I like it. /me changing it

File: ciao-controller/openstack_image.go [hunk 6]
```diff
@@ -50,23 +52,26 @@ func (is ImageService) CreateImage(req image.CreateImageRequest) (image.DefaultR
 			return image.DefaultResponse{}, image.ErrBadUUID
 		}
 
-		img, err := is.ds.GetImage(id)
-		if err != nil {
+		img, err := is.ds.GetImage(tenantID, id)
+		if err != nil && img != (imageDatastore.Image{}) {
```
Comment (obedmr) [inline ciao-controller/openstack_image.go]: @markdryan good catch, thanks

File: ciao-controller/openstack_image.go (hunk 1 above)
Comment (obedmr) [inline ciao-controller/openstack_image.go]: sure, /me fixing it. thanks

File: ciao-image/datastore/datastore_test.go [hunk 7]
```diff
@@ -28,11 +29,60 @@ var metaDsTables = []string{"images"}
...
+func createTmpFile(t *testing.T) *os.File {
+	content := []byte("temporary file's content")
+	tmpfile, err := ioutil.TempFile("", "example")
+	if err != nil {
+		t.Fatal(err)
+		return nil
+	}
+
+	if _, err := tmpfile.Write(content); err != nil {
+		t.Fatal(err)
```
Comment (obedmr) [inline ciao-image/datastore/datastore_test.go]: yeah, good catch, fixed

File: ciao-image/datastore/metads.go (hunk 2 above)
Comment (obedmr) [inline ciao-image/datastore/metads.go]: I added an alias ``imageAPI`` for image package. But, yeah, there's no need of that alias. /me changed it.

File: ciao-image/datastore/metads.go (hunk 3 above)
Comment (obedmr) [inline ciao-image/datastore/metads.go]: same approach as the previous comment. thanks

File: ciao-cli/image.go (hunk 5 above)
Comment (obedmr) [inline ciao-cli/image.go]: done

File: ciao-controller/openstack_image.go [hunk 8]
```diff
@@ -114,16 +119,23 @@ func createImageResponse(img imageDatastore.Image) (image.DefaultResponse, error
...
 
-	images, err := is.ds.GetAllImages()
+	images, err := is.ds.GetAllImages(tenant)
 	if err != nil {
-		glog.Errorf("Error on listing images: %v", err)
+		glog.Errorf("Error on retrieving images from tenant [%v]: %v", tenant, err)
 		return response, err
 	}
 
+	publicImages, err := is.ds.GetAllImages(string(image.Public))
```
Comment (obedmr) [inline ciao-controller/openstack_image.go]: yeah, actually, in database, there's a bucket called ``public`` which only contains public images. I'm thinking that I should also create another one for ``internal`` images, in order to make ``internal`` images retrieval faster.

**Comment by obedmr** [comment]
@markdryan, I'm working on the missing coments, will update soon.

**Comment by tpepper** [comment]
I'd like to see associated with this PR is a refresh of https://github.com/01org/ciao/wiki/Storage.

**Comment by obedmr** [comment]
@markdryan answering your questions:

_In your comments above you say that they can only be created by admin. 
However, I was able to create one in single VM as the normal user._

**This is WIP, we'll be validating admin-privileges in order to create/delete public and internal images. And yes, only admins will be able to create and delete public images.**

_Also who can delete public images? Currently, it only seems to be the user 
that created the images. Admin can see public images but he cannot delete them. 
If he tries he gets a weird 404 error._

**Fixing it, it requires a tenant validation for delete process**

**Comment by obedmr** [comment]
@markdryan  BAT, SingleVM and Deployment changes would be addressed on #1048, #1093 and #1225

**Comment by obedmr** [comment]
@markdryan, current BAT shouldn't fail, it has been fixed. /me testing with singlevm.

**Comment by obedmr** [comment]
singlevm is working.

**Comment by obedmr** [comment]
@markdryan @kaccardi @rbradford  when have a change, could you take a look on this one?

**Comment by markdryan** [comment]
@obedmr SingleVM and BAT are working for me now.  I did notice one odd thing though.

When I do a ciao-cli image list on singlevm it returns nothing.  I know the images are there as I can launch instances.   Is this expected?

**Comment by obedmr** [comment]
@markdryan, are you loading the [~/local/demo.sh](https://github.com/01org/ciao/blob/master/testutil/singlevm/setup.sh#L60-L74) ? If so, demo user shouldn't have access to admin created images, they were private by default, in order to allow demo user to see this, those images should be public ``--visibility public``. I think that would be the required change in SingleVM with this new tenant-separated images approach.

**Comment by obedmr** [comment]
/me adding the singlevm changes on this PR too.

**Comment by markdryan** [comment]
>@markdryan, are you loading the ~/local/demo.sh ?

Yes exactly.

**Comment by obedmr** [comment]
SingleVM changes were just added on a separate commit.

**Comment by obedmr** [comment]
@markdryan, regarding the fact that you can start instances, it's because  workloads were defined with images that exist in ``admin`` tenant. We'll need a mechanism to validate workloads launching in order to verify that user from certain tenant is allowed to use workload's specified images.

@kaccardi, are workloads also tenant-separated? I mean, is there a concept of public or private workloads, if so, we could make sure that admin can create public workloads and private workloads will be the ones that can be only created in a specific tenant.

I think this can be addressed in a separate PR. @rbradford started an issue #1224  for addressing part of this issue.

**Comment by obedmr** [comment]
@rbradford thanks for adding the quotas code.

**Comment by rbradford** [comment]
Hi @obedmr please take a look at the changes I made for quota stuff and let me know if you're happy.

**Comment by obedmr** [comment]
@rbradford sure, I just saw it and in LGTM. I'm happy with the change. Thanks for the help.

**Pull request merged**